DEV-1181 - DEV-1188 -Allow users to interacts with projects and datasets without API keys (#126)
encord-team/encord-client-python: docs/source/authentication.rst

Authentication

> **************

Reviewer: The big thing that jumped out at me from this file in general was that it's actually logically structured, but not in a way that de-emphasizes the API keys. I.e. the text itself de-emphasizes API keys, but the structure makes them equal with user identity. 

While we're on track to move towards 'service accounts' (that is user accounts with specifically scoped permissions, rather than API keys per project/dataset etc) I think we want to move towards 'hiding' the API Keys. Towards that end I would reformat our with structure more similar to below:
```
 - Types of authentication
   - user authentication via public key (*recommended*)
   - resource authentication via api key (supports legacy flows)
 - User Authentication
   - register your key
   - use your key
 - Resource Authentication
   - get a key
   - use the key
```

And then we can slough off resource keys as we move forward

Author: Great point

Types of authentication
========================

> You can authenticate with |company| on a user basis by registering an :xref:`ssh_public_key_authentication` (i.e. "public key" in this document), or on a specific project or dataset basis by using an |company|-generated API key tied to that resource.

Reviewer 2: We ant to hide the use of API keys I think as we just want to tell people to use SSH keys?

Reviewer 2: I see below you have emphasised the use of ssh keys. That might be good enough. @justin-cord what do you think?

Author: I think we should revisit the decision to completely hide this for now because
* The way we have structured the docs now clients know how to use the SSH key and are encouraged to use it in multiple places. This adds the value that we initially wanted to create already.
* If someone is not a project/dataset/organisation admin, they will still be able to use the API key to do some operations.
* If we want to have fine grained access control we need to do this via API keys for now - that's why we'd want to still support them.

Unless we really think that most clients would still be confused about having the API key access, I'd personally keep the documentation in because of the additional use cases. Seems like we can have more value for less work (of deprecating) that way.

Reviewer: quick comment to say my above comment seemed to mirror Riz's comment, catching up on the above...

Public key authentication
----------------------------

> Using the public key authentication will give you full access to all the capabilities of the SDK.

Reviewer 2: ```suggestion
Public key authentication gives you full access to all the capabilities of the SDK.
```

**This is our recommended default authentication method.**

> You will need to choose this authentication method to interact with the :class:`.EncordUserClient`.

Reviewer 2: ```suggestion
Public key authentication is required to use the :class:`.EncordUserClient`.
```

This authentication method will also allow you to interact with a :ref:`general_concepts:Dataset` or a :ref:`general_concepts:Project` assuming you are either of the following:

* An organisation admin of the organisation this project or dataset is part of
* A project admin or project manager of the project that is being accessed
* A dataset admin of the dataset that is being accessed

API key authentication
----------------------

> You can additionally access a specific project or dataset via the API key authentication.

Reviewer 2: ```suggestion
You can access a specific project or dataset via API key authentication.
```

An individual API key is always tied to one specific project or dataset.

We only recommend you to use this authentication if

> * You do not have the access rights to access a project or dataset

Reviewer 2: ```suggestion
* You do not have user access rights for a project or dataset
```

> * You want to access a project or dataset as part of an automated pipeline without using a user access

Reviewer 2: ```suggestion
* You want to access a project or dataset as part of an automated pipeline without using a specific user account
```

>     Getting an API key from the |platform|.

Author: It just is a substitution

>     Getting an API key from the |platform|.

Author: Ah I think I see what you mean - it is a description to the image.

You can additionally create them programmatically as you can see in these guides:

* Create :ref:`tutorials/projects:API keys` for projects
* Create :ref:`tutorials/datasets:API keys` for datasets

Authenticate with |company|
===========================

Once you have registered a public key or created an API key, you can authenticate your SDK client with |company| and get started with the SDK.

Public key authentication with the SDK
--------------------------------------------------

> If you are using public key authentication, authenticate with |company| by passing the corresponding private key to an :class:`.EncordUserClient`.

Reviewer 2: ```suggestion
Once you have registered your public key, authenticate with |company| by passing the corresponding private key to an :class:`.EncordUserClient`.
```
